### PR TITLE
Support including data from S3

### LINKED ISSUE #83: Support including data from outside user-data
EC2 user-data has a rather low max size limit, and when used with CloudFormation it's restricted further to 4096 characters as that's the max size of a parameter there. An encrypted variable in treehugger YAML consumes 200+ bytes, so with only about 15 encrypted parameters we can hit the limit.

It would be nice if the treehugger YAML could support an include statement to go fetch the data from other sources. The most natural AWS'y one is S3. Then a deploy with Cloudformation can just go dump the treehugger file on S3 before updating the stack with the S3 URL (bucket + key + version id) that treehugger sees and fetches. Since treehugger can be run multiple times, it would be good to cache the fetch from S3 to avoid hitting S3 on every single invocation.

I'm imagining the user-data ending up looking like:

```yaml
treehugger:
  include:
  - s3://mybucket/mykey?versionId=abc123
  SOME_OTHER_VAR: 123
  TREEHUGGER_STAGE: production
```

Then when it runs treehugger would fetch all the included files, and merge them all, before continuing.

### PR DESCRIPTION
**Pull request by adammalinowski**
Fixes #83 

Tested in tox and also uploaded a file to s3 to check it is included as expected when calling `include_remote_yaml_data_or_die`.

### CODE REVIEW

File: tests/test_remote.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+        def mock_fetch(bucket_name, key, version):
+            assert bucket_name == 'bucket'
+            assert key == 'filename.yml'
+            if fetch_counter[0] == 0:
+                assert version is None
+            else:
+                assert version == '7'
+            fetch_counter[0] += 1
+            return "X: y"
```
Comment (adamchainz) [inline tests/test_remote.py]: use botocore stub mock like we do for KMS already

**Comment by adamchainz** [comment]
FYI you have to put 'fixes #x' in the PR description not the title

File: treehugger/remote.py [hunk 2]
```diff
@@ -1,19 +1,22 @@
...
-import botocore
+from botocore.exceptions import ClientError
 
 from six.moves.urllib.parse import parse_qs, urlparse
 
 from . import yaml
 from .messaging import die
 
 
+# initialise client here so can be stubbed for testing
```
Comment (adamchainz) [inline treehugger/remote.py]: unnecessary comment, remove

File: tests/test_remote.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            },
+            service_response={
+                'Body': io.BytesIO(b'Z: zzz'),
+            }
+        )
+
+        input_yaml_data = {
+            'A': 'b',
+            'include': [
+                's3://bucket/filename.yml',
```
Comment (adamchainz) [inline tests/test_remote.py]: actually now I think about it, we should reject URL's without a `versionId`. Using `null` as the version might actually fail on S3 buckets without versioning turned on, and anyway versioning is probably the only sane way here.

File: treehugger/remote.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+
+def include_remote_yaml_data_or_die(data):
+    """
+    Check for "include" key with URL(s), if found fetch & include the yaml
+    """
+    include_key = 'include'
+    if include_key in data:
+        url_data = data.pop(include_key)
+        # accept either a single url as value, or list of url values
```
Comment (adamchainz) [inline treehugger/remote.py]: only accept a list, simpler

File: treehugger/remote.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        if not isinstance(url_data, list):
+            url_data = [url_data]
+        for url in url_data:
+            parsed_url = urlparse(url)
+            if parsed_url.scheme.lower() == 's3':
```
Comment (adamchainz) [inline treehugger/remote.py]: simpler code structure for cases like this:

```py
if not X:
    error_case()
```

then continue with the happy path. Avoids unnecessary indentation

File: HISTORY.rst [hunk 6]
```diff
@@ -9,6 +9,12 @@ Pending Release
 
 * (Insert new release notes below this line)
 
+2.2.0 (2018-01-17)
+------------------
```
Comment (adamchainz) [inline HISTORY.rst]: don't insert version numbers, that's done at release time https://github.com/timeoutdigital/treehugger/commit/dd721cb75c7fd50490f8b69a2c2527cc8e79274c

File: treehugger/__init__.py [hunk 7]
```diff
@@ -1,4 +1,4 @@
 # -*- coding:utf-8 -*-
 from __future__ import absolute_import, division, print_function, unicode_literals
 
-__version__ = '2.1.0'
+__version__ = '2.2.0'
```
Comment (adamchainz) [inline treehugger/__init__.py]: don't bump version number, as above

**Comment by adamchainz** [comment]
Please attend to previous review.

Also this doesn't work E2E because `EnvironmentDict.from_yaml_dict` is checking the schema of the treehugger file and the 'include' value is a list which isn't allowed, error from line 38 of data.py. Please update.

Also probably a good idea to add a more E2E test in `test_cli.yml`, could just be testing 'print' with a file that points to include - cloning `test_print_file` and extending it?

**Comment by adamchainz** [comment]
Transcript of my testing session:

```
$ aws s3api create-bucket --bucket treehugger-pr-91 --create-bucket-configuration LocationConstraint=eu-west-1
{
    "Location": "http://treehugger-pr-91.s3.amazonaws.com/"
}
$ aws s3api put-bucket-versioning --bucket treehugger-pr-91 --versioning-configuration Status=Enabled,MFADelete=Disabled
$ echo 'TREEHUGGER_APP: foo
TREEHUGGER_STAGE: bar
MY_UNENCRYPTED_VAR: hello
MY_ENCRYPTED_VAR: {to_encrypt: world}' > test.yml
$ python3 -m treehugger encrypt-file test.yml
Successfully encrypted
$ cat test.yml
MY_ENCRYPTED_VAR: {encrypted: AQICAHjIWrbIlXQm7WWrYJc6hj7L2ibsvmI6sywFSdkX0xuxtwHHnteTDzAoceFHpEV9Hx5xAAAAYzBhBgkqhkiG9w0BBwagVDBSAgEAME0GCSqGSIb3DQEHATAeBglghkgBZQMEAS4wEQQMC/DHjQtmXfaCjvd6AgEQgCDE4Fxgdzs14kA0WIOCplp5TUoIfZGaVfuffWwoWi13kw==}
MY_UNENCRYPTED_VAR: hello
TREEHUGGER_APP: foo
TREEHUGGER_STAGE: bar
$ aws s3 cp test.yml s3://treehugger-pr-91/
upload: ./test.yml to s3://treehugger-pr-91/test.yml
$ aws s3api head-object --bucket treehugger-pr-91 --key test.yml
{
    "AcceptRanges": "bytes",
    "ContentType": "binary/octet-stream",
    "LastModified": "Wed, 24 Jan 2018 11:38:27 GMT",
    "ContentLength": 312,
    "VersionId": "JqLIeiip6r3N.bxltXGk7OzVoI88o64a",
    "ETag": "\"b43a194a175606fd1f00b2642451b3f1\"",
    "Metadata": {}
}
$ echo 'include: ["s3://treehugger-pr-91/test.yml?versionId=JqLIeiip6r3N.bxltXGk7OzVoI88o64a"]' > test2.yml
$ python3 -m treehugger print -f test2.yml
...
ValueError: Invalid object in treehugger data: ['s3://treehugger-pr-91/test.yml?versionId=JqLIeiip6r3N.bxltXGk7OzVoI88o64a']
```

**Comment by adammalinowski** [comment]
Yes will make an E2E test, good idea.

Rather than making `from_yaml_dict` handle lists or a special case for `include`, would it make more sense for `include` to just take a single value? I can't think why you'd need to include more than one file.

**Comment by adamchainz** [comment]
I guess for now we can take a single value, we can always extend it by allowing lists there in future if we do come up with a multiple file include situation.

File: README.rst [hunk 8]
```diff
@@ -97,10 +97,9 @@ You can also pull in data from external sources by using the key `include`, for
 
 .. code-block:: yaml
 
-    include: 
-    - s3://bucket-name/filename.yml?versionId=7
+    include: s3://bucket-name/filename.yml?versionId=7
 
-Where the value is a list of urls pointing to other yaml files consisting of environment variables. The files will be fetched, and the yaml data included. Currently only S3 urls are supported. versionId is required.
+Where the value is a url pointing to another yaml file consisting of environment variables. The file will be fetched, and the yaml data included. Currently only S3 urls are supported. versionId is required.
```
Comment (adamchainz) [inline README.rst]: URL
YAML

Acronyms are capitalized

File: tests/test_cli.py [hunk 9]
```diff
@@ -461,6 +462,58 @@ def test_print_file(self, tmpdir, kms_stub, capsys):
...
+        )
+        s3_stub.add_response(
+            'get_object',
+            expected_params={
+                'Bucket': 'my-bucket',
+                'Key': 'my_file.yml',
+                'VersionId': '2',
+            },
+            service_response={
+                'Body': io.BytesIO(b'MY_INCLUDED_VAR: IliveinAWS'),
```
Comment (adamchainz) [inline tests/test_cli.py]: 'IliveonS3' would be more accurate, everything is on AWS here in some way

**Comment by adamchainz** [comment]
Works for me! Following on from my test session:

```
$ echo 'include: s3://treehugger-pr-91/test.yml?versionId=JqLIeiip6r3N.bxltXGk7OzVoI88o64a' > test2.yml
$ python3 -m treehugger print -f test2.yml
MY_ENCRYPTED_VAR=world
MY_UNENCRYPTED_VAR=hello
TREEHUGGER_APP=foo
TREEHUGGER_STAGE=bar
```

**Pull request merged**